### DISCUSSION TITLE
How to use multiple database with Django and SQLAlchemy

### DISCUSSION THREAD

**Discussion by akashchevli** [sqlalchemy/sqlalchemy#11038] [Usage Questions] [answered]
**Description**:

I have a Django application that currently stores data for multiple brands in a single database. Each brand's data, including user details and relevant information, is stored in separate tables within this database. However, I now have a requirement to migrate the application to use separate databases for each brand.

**Current Setup**:

One database containing tables for multiple brands. User details and relevant data for each brand stored in separate tables within this database.

**Desired Setup**:

Separate databases for each brand. User details and relevant data for each brand stored in their respective databases.

**Tools Used**:

Django for web framework. SQLAlchemy for database ORM. Alembic for database migrations.

**Key Questions**:

1. How can I configure Django to use separate databases for each brand?
2. What changes do I need to make to SQLAlchemy models to support this new database structure?
3. How should I handle database migrations using Alembic to ensure seamless transition to the new setup?
4. Are there any best practices or potential pitfalls I should be aware of during this migration process?
5. As of now, I'm storing different brands's user information in one table. Now, I need to separate the database for each brand, but the table schema will remain the same for all brands. So if I alter the column or add a new column to any of the brand database tables, how can I ensure that that column also gets updated in all brand database tables as well?

Currently, I'm following the simple technique to connect the single Database.
```
engine = create_engine(<Database URL>)
session = scoped_session(sessionmaker(bind=engine, autocommit=True))
Base.metadata.create_all(engine)
```

Any guidance, examples, or resources would be greatly appreciated. Thank you!

**Comment by CaselIT** [comment]
Hi,

You are talking about multiple databases, but it seems that are really referring to multiple schemas.
Can you clarify which architecture are you thinking about? Usually multiple databases require multiple url, while multiple schemas share the same url

**Reply by akashchevli** [reply]
Hi @CaselIT, I appreciate your initiative in providing a response.

I'm discussing about multiple databases.

The current flow is that, I have one database, and one database contains multiple tables.
Desire flow: I want to create a separate database for each brand, and each brand has the same tables and schemas.

So let's say there is one brand, MacD.
I want to create a separate database for MacD, and all the users that belong to MacD will be stored in the users table of the MacD database.
Same for the brand, Domino's, 
There is a separate database for Domino's, and all the users that belong to Domino's will be stored in the users table of Domino's database.

Here, the schema of the MacD and Domino's is the same, but both tables are present in both databases.
Just to note, I want to do in mid project.

**Reply by CaselIT** [reply]
At the moment the only support for this is to use different engines in sqlalchemy. You will have to manage them manually.

For alembic you can probably either use a different url using an env variable or similar or use a custom argument to select the url given a static list. See https://alembic.sqlalchemy.org/en/latest/api/runtime.html#alembic.runtime.environment.EnvironmentContext.get_x_argument

My question regarding db vs schema is that in case of schema you could have used something like this https://docs.sqlalchemy.org/en/20/core/connections.html#schema-translating

Don't know what kind of changes you will need to do to django, sorry

**Reply by akashchevli** [reply]
Okay, if I create different engines for different brands, let's say I add a new column in one of the brands (i.e., MacD), then that column also gets added in another brand (i.e., Domino's) when I create a migration file using the alembic command? Does that column also appear in Domino's engine as well?

**Reply by CaselIT** [reply]
How are you adding the column? In the db directly using alter column?

**Reply by akashchevli** [reply]
No, after adding the column to the model, I'll perform the `alembic revision -m "Add new column to users table."` Thus, the new migration file will be created by this command. I now want the newly created column for each engine (MacD engine, Domino's engine, etc.) in the migrations file (Of course, inside the upgrade_<engine_name> function). Therefore, the modification will be reflected on that specific table across all databases when I run `alembic upgrade head`.

**Reply by CaselIT** [reply, marked as answer]
You will have to modify the alembic env file to apply each command to every engine. You can probably do something like the multidb template, but using a single metadata https://github.com/sqlalchemy/alembic/blob/main/alembic/templates/multidb/env.py

**Reply by akashchevli** [reply]
Useful. Thanks for the help.
